docs/development_best_practices.md

```diff
@@ -0,0 +1,130 @@
+2. An optional **statistics (stats) listener** that specifies *when* to partition. The reason it is optional is that it is needed only when dynamic partitioning is needed. With the stats listener, the stats can be used to determine when to partition.
+3. In some cases the *operator* itself should be aware of partitioning and would need to provide supporting code.
+	* In case of input operators each partition should have a property or a set of properties that allow it to distinguish itself from the other partitions and fetch unique data.
+4. When an operator that was originally a single instance is split into multiple partitions with each partition working on a subset of data, the results of the partitions may need to be combined together to compute the final result. The combining logic would depend on the logic of the operator. This would be specified by the developer using a **Unifier**, which is deployed as another operator by the platform. If no *Unifier* is specified, the platform inserts a **default unifier** that merges the results of the multiple partition streams into a single stream. Each output port can have a different *Unifier* and this is specified by returning the corresponding *Unifier* in the **getUnifier** method of the output port. The operator developer should provide a custom *Unifier* wherever applicable.
+5. The Apex *engine* that brings everything together and effects the partitioning.
+
+Since partitioning is critical for scalability of applications, operators must support it. There should be a strong reason for an operator to not support partitioning, such as, the logic performed by the operator not lending itself to parallelism. In order to support partitioning, an operator developer, apart from developing the functionality of the operator, may also need to provide a partitioner, stats listener and supporting code in the operator as described in the steps above. The next sections delve into this. 
+
+### Out of the box partitioning
+
+Apex comes with some built-in partitioning utilities that can be used in certain scenarios.
+
+* **StatelessPartitioner** in Malhar provides a default partitioner, which can be used in certain conditions. If the operator satisfies these conditions, no special partitioning code is needed . These are
+No dynamic partitioning is needed, see next point about dynamic partitioning. 
+There is no distinct initial state for the partitions, i.e., all partitions start with the same initial state submitted during application launch. Typically input or output operators will not fall into this category, although there are exceptions and mainly it is operators in the middle of the DAG that are in this category.
+When used with non-input operators, only the data for the first declared input port is distributed among the different partitions. All other input ports are treated as broadcast and all partitions receive all the data for that port.
```

Comment:
This doesn't fit the pattern of being a "condition" that the operator satisfies (the two earlier points do). Suggest rewording.